### PR TITLE
Feature: export to pdf

### PR DESCRIPTION
**Pull request by wolfderechter**
## ID

DGS-306

## Description

Generate pdf for subsidy forms

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Other

## How to test

Open subsidies and click 'Download huidige stap (PDF)' to verify working and valid pdf's of the subsidy step.

### CODE REVIEW

File: app/styles/print.scss [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    flex-direction: column-reverse; // avoid table collapsing with next element
+  }
+  .au-c-app {
+    // overflow: visible;
+    height: auto;
+  }
+
+  .hide-on-print {
+    display: none;
+  }
```
Comment (Windvis) [inline app/styles/print.scss:16]: Appuniversum already ships this one as `.au-u-hide-on-print`

Comment (wolfderechter) [reply app/styles/print.scss:16]: ooooh okay I'll use that one

File: app/controllers/subsidy/applications/edit.js [hunk 2]
```diff
@@ -2,9 +2,18 @@ import Controller from '@ember/controller';
...
+import { saveAs } from 'file-saver';
+import JSZip from 'jszip';
 
 export default class SubsidyApplicationsEditController extends Controller {
   @service router;
+  @service() store;
+
+  constructor() {
+    super(...arguments);
+    window.addEventListener('beforeprint', this.prepareTextareasForPrinting);
```
Comment (Windvis) [inline app/controllers/subsidy/applications/edit.js:15]: Doesn't this event handler cause issues if they would try to print a different page in the app?

The handler assumes the user is on the edit page, but after visiting the edit page and going to a different route it would still be active AFAIK. I think it might be possible to start / stop the handler from the route's setupController / resetController hooks. Alternatively you could add a new `PrintHandler` component that receives the method as an argument, and do the registration there, and tear it down again in the `willDestroy` method. A custom modifier would also work. 

Multiple options, but I think removing the handler somehow would be the safest?

File: app/controllers/subsidy/applications/edit.js [hunk 3]
```diff
@@ -28,6 +37,95 @@ export default class SubsidyApplicationsEditController extends Controller {
   get canDelete() {
     return this.model.consumption.get('status.isConcept');
   }
+  @action
+  async exportSubsidyAsPDF() {
+    const previousDocumentTitle = document.title;
+    const filename = `${await this.createFilename()}.pdf`;
+
+    document.title = filename;
```
Comment (Windvis) [inline app/controllers/subsidy/applications/edit.js:45]: Why is this needed exactly? What does the browser do with this? It might be worth adding a comment about that here. It feels strange without context.

File: app/styles/print.scss [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  .au-c-main-header__title--link {
+    overflow: visible;
+  }
+  // Hides the title href
+  .au-c-main-header__title--link:after {
+    content: none !important;
+  }
+
+  // show the logged in user fully
+  .au-c-dropdown__name[data-test-dropdown-title] {
```
Comment (Windvis) [inline app/styles/print.scss:37]: I wouldn't use those `data-test` attributes. Can't you simply add a new css class instead?

File: app/styles/print.scss [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    max-width: fit-content;
+  }
+
+  .au-c-list-navigation__link:after {
+    content: none !important;
+  }
+
+  .subsidySteps {
```
Comment (Windvis) [inline app/styles/print.scss:45]: Nitpick, but we use dasherized / BEM names for custom css classes.

File: app/controllers/subsidy/applications/edit.js [hunk 6]
```diff
@@ -28,6 +37,95 @@ export default class SubsidyApplicationsEditController extends Controller {
...
+    const currentStep = await this.store.findRecord(
+      'subsidy-application-flow-step',
+      currentStepID,
+      {
+        include: 'subsidy-procedural-step',
+      }
+    );
+    const currentProceduralStep = await currentStep.subsidyProceduralStep;
+    const currentProceduralStepName = currentProceduralStep.description;
+    const currentSubsidyName = this.consumption.get(
```
Comment (Windvis) [inline app/controllers/subsidy/applications/edit.js:104]: I wouldn't use `.get` anymore. Just await the relationship first, as you're doing in the other code already.

File: app/templates/subsidy/applications/edit.hbs [hunk 7]
```diff
@@ -17,17 +17,35 @@
...
+        role="menu"
+        class="au-u-hide-on-print"
+      >
+        <AuButton @skin="link" role="menuitem" {{on "click" this.exportSubsidyAsPDF}}>
+          Download huidige stap (PDF)
+        </AuButton>
+        <AuButton @skin="link" role="menuitem" {{on "click" this.downloadBijlagen}}>
+          Alle bijlagen downloaden in .zip
+        </AuButton>
+      </AuDropdown>
```
Comment (Windvis) [inline app/templates/subsidy/applications/edit.hbs:36]: It feels a bit strange to always show this, even if the form hasn't been submitted yet?

File: app/controllers/subsidy/applications/edit.js [hunk 8]
```diff
@@ -2,9 +2,18 @@ import Controller from '@ember/controller';
 import { inject as service } from '@ember/service';
 import { task } from 'ember-concurrency';
 import fetch from 'fetch';
+import { action } from '@ember/object';
+import { saveAs } from 'file-saver';
+import JSZip from 'jszip';
```
[the remark this answers is not in the document]
Comment (Windvis) [reply app/controllers/subsidy/applications/edit.js:7]: But related to that. It seems to cover the same function as the PR I linked, why reimplement that here?

File: app/styles/print.scss
Comment (Windvis) [inline app/styles/print.scss:1]: Hmm, this whole CSS-based printing setup feels very frail and prone to regression issues in the future. Every change we make to the frontend has the potential to break the print styles.

I wonder if there isn't a more solid approach to this, and assume this is only a temporary solution instead.

File: app/controllers/subsidy/applications/edit.js (hunk 6 above)
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js:104]: I got a warning when I didn't use get

File: app/controllers/subsidy/applications/edit.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js:7]: I believe the PR's functionality works on a per file-upload input base, in here we wanted to combine all the files in from the form in a zip. I tried re-using some of the other code with client-zip but that seemed more complex so I just kept this, but I can take another look

File: app/controllers/subsidy/applications/edit.js (hunk 3 above)
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js:45]: You're right, will add a comment. It's a trick to use the subsidy and step name in the pdf.

File: app/controllers/subsidy/applications/edit.js (hunk 6 above)
Comment (Windvis) [reply app/controllers/subsidy/applications/edit.js:104]: You need to await the relationship before accessing properties. Similar to the `currentStep` code above it.

File: app/styles/print.scss (hunk 4 above)
Comment (wolfderechter) [reply app/styles/print.scss:37]: I could but I'd have to add it in https://github.com/lblod/ember-submission-form-fields

File: app/controllers/subsidy/applications/edit.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Windvis) [reply app/controllers/subsidy/applications/edit.js:7]: But how often does it happen that there are multiple different file upload fields in the same form?

We opted out of this in ember-submission-form-fields because it made things more complex for little value. Most forms (in toezicht) have only a single file upload and/or remote url field, so this would have saved 1 click. Not sure about the subsidy forms, but still.

File: app/styles/print.scss
Comment (wolfderechter) [reply app/styles/print.scss:1]: Agree but hard to beat, any other solution would require a lot more time. At least the css printing is flexible and easy to update. Given that we have limited time to work on this feature, and the main focus is the frontend.

But ultimately any pdf printing solution will be hard since subsidies have a lot of wide and long tables that don't go well together with a4 pdf's. So users should still primarily use the frontend.

In the future we can for sure revisit the functionality and try to upgrade to a more resilient option!

File: app/templates/subsidy/applications/edit.hbs (hunk 7 above)
Comment (wolfderechter) [reply app/templates/subsidy/applications/edit.hbs:36]: Agreed, I'll hide it when the subsidystep is in concept phase

File: app/controllers/subsidy/applications/edit.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js:7]: For subsidies it happens quite often actually, fietsinfrastructuur for example has 2 steps with 5+ file uploads. Plus it's handy to have the button at the top of the page

[the remark this answers is not in the document]
Comment (Windvis) [reply app/controllers/subsidy/applications/edit.js:7]: In any case, I would try to split off the batch download feature from the pdf printing work. It feels unrelated and I think we can come up with a better solution.

File: app/styles/print.scss (hunk 4 above)
Comment (Windvis) [reply app/styles/print.scss:37]: But this is custom code from the app, no? I don't think this is a form-field component?

File: app/controllers/subsidy/applications/edit.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Windvis) [reply app/controllers/subsidy/applications/edit.js:7]: > For subsidies it happens quite often actually, fietsinfrastructuur for example has 2 steps with 5+ file uploads. Plus it's handy to have the button at the top of the page

Ah, I see, I missed this reply. Yea that makes sense then. I still don't like the way it's implemented though :sweat_smile: . Depending on implementation details of the forms addon feels very frail.

File: app/styles/print.scss
Comment (Windvis) [reply app/styles/print.scss:1]: Yea, I'm not opposed to the CSS styles themselves, but the selectors seem to depend on implementation details of code we don't maintain in the app, so every update of the forms addon potentially breaks the printing styles. I simply don't like it, but maybe it's good enough for now.

File: app/controllers/subsidy/applications/edit.js (hunk 2 above)
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js]: Good catch, the current implementation also kept duplicating when clicking the print multiple times. I added `{ once: true }` and put the addeventlistener of afterprint before calling the window.print(). This should fix it

File: app/styles/print.scss (hunk 4 above)
Comment (wolfderechter) [reply app/styles/print.scss:37]: Sorry I was confusing it with something else, you're right I can just add a css class

File: app/controllers/subsidy/applications/edit.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js]: I refactored to use client-zip with triggerZipDownload instead of jszip and file-saver. I hope this will do and we can merge it in together :)

File: app/controllers/subsidy/applications/edit.js (hunk 2 above)
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js]: I see! How about we just call the prepare function inside the exportSubsidyAsPDF function?
```

  @action
  async exportSubsidyAsPDF() {
    await this.prepareTextareasForPrinting();
    ...
 ```
 This makes things a lot easier, and we can assume the user will have to click the button anyway to have the right filename.

Comment (Windvis) [reply app/controllers/subsidy/applications/edit.js]: Yea, I think that makes sense!

File: app/controllers/subsidy/applications/edit.js [hunk 9]
```diff
@@ -29,6 +33,107 @@ export default class SubsidyApplicationsEditController extends Controller {
     return this.model.consumption.get('status.isConcept');
   }
 
+  get canDownload() {
+    return !this.model.consumption.get('status.isConcept');
```
Comment (Windvis) [inline app/controllers/subsidy/applications/edit.js:37]: Not 100% sure if this is the correct check. In my head it only makes sense to print a form that is in read-only mode. Not sure how hard that would be, and it's already better than always showing it at least.

For example, the button is displayed, but the form can still be edited:
image

Opinions?

File: app/templates/subsidy/applications/edit.hbs [hunk 10]
```diff
@@ -17,17 +17,37 @@
...
+          @skin="secondary"
+          @icon="download"
+          @iconAlignment="left"
+          role="menu"
+          class="au-u-hide-on-print"
+        >
+          <AuButton @skin="link" role="menuitem" {{on "click" this.exportSubsidyAsPDF}}>
+            Download huidige stap (PDF)
+          </AuButton>
+          <AuButton @skin="link" role="menuitem" {{on "click" this.downloadBijlagen}}>
```
Comment (Windvis) [inline app/templates/subsidy/applications/edit.hbs:34]: Another strange thing I noticed. The button is visible even if there are no download links, so if the user clicks it nothing happens.

Probably an edge case, and fixing it might be tricky since it depends on the form being rendered first so maybe not worth putting time in it?

Comment (wolfderechter) [reply app/templates/subsidy/applications/edit.hbs]: Good idea! Just pushed a simple check for it

File: app/controllers/subsidy/applications/edit.js (hunk 9 above)
Comment (wolfderechter) [reply app/controllers/subsidy/applications/edit.js:39]: I checked with Marie and we want to always show the button

**Review by Windvis — APPROVED** [review]

File: app/templates/subsidy/applications/edit.hbs [hunk 11]
```diff
@@ -120,6 +120,13 @@
     </nav>
   </:sidebar>
   <:content>
+    <p class="display-on-print au-u-text-center au-u-margin-top">
+      {{#if this.stepIsSubmitted}}
+        Dit formulier is ingediend
+      {{else}}
+        Dit formulier is in concept fase
+      {{/if}}
+    </p>
```
Comment (wolfderechter) [inline app/templates/subsidy/applications/edit.hbs:129]: @Windvis I have added this clarification above the form in the pdf
